src/marketplace/PichiMarketplace.sol

```diff
@@ -0,0 +1,296 @@
+        external
+        initializer
+    {
+        if (weth_ == address(0) || marketplaceFeeRecipient_ == address(0)) revert InvalidAddress();
+        if (precision_ == 0) revert InvalidValue();
+        domainSeparator = keccak256(
+            abi.encode(
+                keccak256("EIP712Domain(string name,string version,uint256 chainId, address verifyingContract)"),
+                keccak256("PichiMarketplace"),
+                keccak256(bytes("1")),
+                block.chainid,
+                address(this)
+            )
+        );
+        weth = weth_;
+        marketplaceFeeRecipient = marketplaceFeeRecipient_;
+        marketplaceFee = marketplaceFee_;
+        precision = precision_;
+
+        __Ownable_init();
+    }
+
+    /// @notice Cancels all orders for a user by setting their userMinOrderNonce to minNonce specified
+    /// @param minNonce Minimum nonce for orders that can be executed
+    function cancelAllOrdersForCaller(uint256 minNonce) external override {
+        if (minNonce <= userMinOrderNonce[msg.sender]) revert NonceLowerThanCurrent();
+
+        userMinOrderNonce[msg.sender] = minNonce;
+
+        emit AllOrdersCancelled(msg.sender, minNonce);
+    }
+
+    /// @notice Cancels specific order nonces for a user
+    /// @param orderNonces Array of order nonces to mark as cancelled
+    function cancelOrdersForCaller(uint256[] calldata orderNonces) external override {
+        if (orderNonces.length == 0) revert ArrayEmpty();
+
+        for (uint256 i = 0; i < orderNonces.length; i++) {
+            if (orderNonces[i] <= userMinOrderNonce[msg.sender]) revert NonceLowerThanCurrent();
+            if (isUserNonceExecutedOrCancelled[msg.sender][orderNonces[i]]) revert OrderAlreadyCancelled();
+            isUserNonceExecutedOrCancelled[msg.sender][orderNonces[i]] = true;
+        }
+
+        emit OrdersCancelled(msg.sender, orderNonces);
+    }
+
+    /// @notice Called by buyer to execute a seller's listing
+    /// @notice Listing must be for a collection that is accepted that is paid in an accepted currency
+    /// @param listing Listing order data
+    function executeListing(Listing calldata listing) external payable override {
+        if (!isCurrencyAccepted[listing.order.currency]) revert CurrencyNotAccepted();
+        if (!isCollectionAccepted[listing.order.collection]) revert CollectionNotAccepted();
+        if (msg.sender == listing.seller) revert OrderCreatorCannotExecute();
+
+        _validateListing(listing);
+
+        if (msg.value != 0) {
+            if (listing.order.currency != weth) revert CurrencyMismatch();
+            _transferWalletForPayment(listing.order, listing.seller, msg.sender, true);
+        } else {
+            _transferWalletForPayment(listing.order, listing.seller, msg.sender, false);
+        }
+
+        isUserNonceExecutedOrCancelled[listing.seller][listing.nonce] = true;
```

Comment:
Is there a re-entrancy attack possible here?

Reply:
Should be set to true before transferring payments so that _validateListing() fails on re-entry

Reply:
fixed